Sources/MongoSwift/APM.swift

```diff
@@ -129,41 +250,80 @@ public struct CommandFailedEvent: MongoCommandEvent, InitializableFromOpaquePoin
     /// The connection id for the command.
     public let serverAddress: Address
 
-    /// Initializes a CommandFailedEvent from an OpaquePointer to a mongoc_apm_command_failed_t
-    fileprivate init(_ event: OpaquePointer) {
-        self.duration = mongoc_apm_command_failed_get_duration(event)
-        self.commandName = String(cString: mongoc_apm_command_failed_get_command_name(event))
+    fileprivate init(mongocEvent: MongocCommandFailedEvent) {
+        self.duration = mongoc_apm_command_failed_get_duration(mongocEvent.ptr)
+        self.commandName = String(cString: mongoc_apm_command_failed_get_command_name(mongocEvent.ptr))
         var error = bson_error_t()
-        mongoc_apm_command_failed_get_error(event, &error)
-        let reply = Document(copying: mongoc_apm_command_failed_get_reply(event))
+        mongoc_apm_command_failed_get_error(mongocEvent.ptr, &error)
+        let reply = Document(copying: mongoc_apm_command_failed_get_reply(mongocEvent.ptr))
         self.failure = extractMongoError(error: error, reply: reply) // should always return a CommandError
-        self.requestId = mongoc_apm_command_failed_get_request_id(event)
-        self.operationId = mongoc_apm_command_failed_get_operation_id(event)
-        self.serverAddress = Address(mongoc_apm_command_failed_get_host(event))
+        self.requestId = mongoc_apm_command_failed_get_request_id(mongocEvent.ptr)
+        self.operationId = mongoc_apm_command_failed_get_operation_id(mongocEvent.ptr)
+        self.serverAddress = Address(mongoc_apm_command_failed_get_host(mongocEvent.ptr))
     }
-}
 
-/// A protocol for SDAM monitoring events to implement.
-public protocol MongoSDAMEvent: MongoEvent {}
-
-/// A protocol for SDAM monitoring events that relate to topology updates to implement, specifying the id of the
-/// topology being updated.
-public protocol MongoTopologyUpdateEvent: MongoEvent {
-    /// The id of the topology being updated.
-    var topologyId: ObjectId { get }
+    fileprivate func toPublishable() -> CommandEvent {
+        return .failed(self)
+    }
 }
 
-/// A protocol for topology update events that relate to specific servers to implement, specifying the address of the
-/// server being updated.
-public protocol MongoServerUpdateEvent: MongoTopologyUpdateEvent {
-    /// The address of the server being updated.
-    var serverAddress: Address { get }
+/// An SDAM monitoring event related to topology updates.
+public enum TopologyEvent: Publishable {
```

Comment:
I don't have a great sense of how people use SDAM monitoring so maybe it wouldn't be that useful, but should we consider not grouping together the topology + server events into one type?

Reply:
yea, hmmm. the one plus I can think of for splitting them up is that it would make things marginally easier if you only care about topology or server but not both.
 
I believe the server events are redundant with the topology events since any server change also results in a topology change, and the topology descriptions contain server descriptions. so it seems plausible to me that you would only want topology events.